File: src/main/clojure/comzotohlabscljc/netty/expect100.clj
;; This library is distributed in  the hope that it will be useful but without
;; any  warranty; without  even  the  implied  warranty of  merchantability or
;; fitness for a particular purpose.
;; The use and distribution terms for this software are covered by the Eclipse
;; Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
;; can be found in the file epl-v10.html at the root of this distribution.
;; By using this software in any  fashion, you are agreeing to be bound by the
;; terms of this license. You  must not remove this notice, or any other, from
;; this software.
;; Copyright (c) 2013-2014 Cherimoia, LLC. All rights reserved.

(ns ^{ :doc ""
       :author "kenl" }

  comzotohlabscljc.netty.expect100

  (:require [clojure.tools.logging :as log :only [info warn error debug] ])
  (:require [clojure.string :as cstr])
  (:import (io.netty.buffer Unpooled))
  (:import (io.netty.channel ChannelHandler ChannelHandlerContext
                             ChannelFutureListener ChannelFuture
                             ChannelPipeline
                             ChannelInboundHandlerAdapter))
  (:import (io.netty.handler.codec.http HttpMessage HttpResponseStatus
                                        HttpHeaders HttpVersion
                                        DefaultFullHttpResponse)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;(set! *warn-on-reflection* false)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(def ^:private CONTINUE (DefaultFullHttpResponse. (HttpVersion/HTTP_1_1)
                                                  (HttpResponseStatus/CONTINUE)
                                                  (Unpooled/EMPTY_BUFFER)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn Expect100 ""

  ^ChannelHandler
  []

  (proxy [ChannelInboundHandlerAdapter] []
    (channelRead [c msg]
      (let [^ChannelHandlerContext ctx c]
        (when (and (instance?  HttpMessage msg)
                   (HttpHeaders/is100ContinueExpected msg))
          (-> (.writeAndFlush ctx CONTINUE)
              (.addListener (reify ChannelFutureListener
                              (operationComplete [_  f]
                                (when-not (.isSuccess ^ChannelFuture f)
                                          (.fireExceptionCaught ctx (.cause f))))))))
        (.fireChannelRead ctx msg)

      ))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn AddExpect100 ""

  ^ChannelPipeline
  [pipe options]

  (let []
    (.addLast ^ChannelPipeline  pipe "e100" (Expect100))
    pipe
  ))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(def ^:private expect100-eof nil)

File: src/main/java/com/zotohlabs/frwk/netty/Expect100.java
/*??
*
* Copyright (c) 2013 Cherimoia, LLC. All rights reserved.
*
* This library is distributed in the hope that it will be useful
* but without any warranty; without even the implied warranty of
* merchantability or fitness for a particular purpose.
*
* The use and distribution terms for this software are covered by the
* Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
* which can be found in the file epl-v10.html at the root of this distribution.
*
* By using this software in any fashion, you are agreeing to be bound by
* the terms of this license.
* You must not remove this notice, or any other, from this software.
*
 ??*/

package com.zotohlabs.frwk.netty;

import io.netty.channel.*;
import io.netty.handler.codec.http.*;

/**
 * @author kenl
 */
@ChannelHandler.Sharable
public class Expect100 extends SimpleChannelInboundHandler {

  private static final Expect100 shared = new Expect100();
  public static Expect100 getInstance() {
    return shared;
  }

  private static HttpResponse c100_rsp() {
    return new DefaultFullHttpResponse(HttpVersion.HTTP_1_1,
                                                  HttpResponseStatus.CONTINUE);
  }

  public static void handle100( ChannelHandlerContext ctx, HttpMessage msg) {
    if (HttpHeaders.is100ContinueExpected(msg)) {
      ctx.writeAndFlush( c100_rsp() ).addListener(new ChannelFutureListener() {
        public void operationComplete(ChannelFuture f) {
          if (!f.isSuccess()) {
            ctx.fireExceptionCaught(f.cause());
          }
        }
      });
    }
  }

  @Override
  protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Exception {
    if (msg instanceof HttpMessage) {
      handle100(ctx, (HttpMessage) msg);
    }
    ctx.fireChannelRead(msg);
  }

}


